### DISCUSSION TITLE
How to migrate next@8.1.0 to next@9 ?

### DISCUSSION THREAD

**Discussion by dhavalsoni2001** [vercel/next.js#12005] [Help] [answered]
I have an application on next@8.1.0, I am planning to migrate it to next@9. I have backend and frontend code on nextjs framework. So how difficult to migrate nextjs 8 to nextjs 9? 

Is there any documentation that helps out to migration of nextjs 8 to nextjs 9?

I went through this documentation but then API call was throwing `404 `error.  The same API is working in nextjs 8.

https://nextjs.org/docs/upgrading 

API URL : `http://localhost:3000/api/getUserData` 

I have also seen that in nextjs 9 we can create `api` folder in `pages`.

**Comment by timneutkens** [comment]
Sounds like you might be using a custom server and that the server is not implemented correctly where it hands off the request to Next.js before matching any other routes you've added. Can you share the custom server?

**Comment by dhavalsoni2001** [comment]
Yes that's right. I am using the express.

```
import { addUser } from "./controllers/user";

...

app
  .prepare()
  .then(() => {
    const server = express();
    server.use(bodyParser.json());
    server.use(bodyParser.urlencoded({ extended: true }));

    server.post("/api/addUser",validateAuthentication,addUser);`

**Reply by timneutkens** [reply]
Can you post the full server?

**Reply by dhavalsoni2001** [reply]
Hey @timneutkens Sorry for the delayed response... 

package.json

```  
"scripts": {
    "dev": "node start-server.js",
    "build": "next build",
    "start": "SET NODE_ENV=production & node  start-server.js"
  },
```

start-server.js

```
require('babel-register')({
    presets: [ 'es2015' ]
 });
 
 require('./server');
```

server/index.js
```
const express = require("express");
const next = require("next");
const mongoose = require("mongoose");
const dev = process.env.NODE_ENV !== "production";
const app = next({ dev });
const bodyParser = require("body-parser");
const handle = app.getRequestHandler();
const dotenv = require("dotenv");

import { getUsers } from "./controllers/user";
i

import { validateAuthentication } from "./services/auth";


dotenv.config();

mongoose.connect("mongodb://127.0.0.1:27017/database", {
  useNewUrlParser: true,
});
mongoose.connection.on(
  "error",
  console.error.bind(console, "connection error:")
);

app
  .prepare()
  .then(() => {
    const server = express();
    server.use(bodyParser.json());
    server.use(bodyParser.urlencoded({ extended: true }));

    server.get("/api/getUsers", validateAuthentication, getUsers);
     // ......
    // All  APIs 
    // ......
    server.use(function (req, res, next) {
      res.header("Access-Control-Allow-Origin", "*");
      res.header(
        "Access-Control-Allow-Headers",
        "Origin, X-Requested-With, Content-Type, Accept"
      );
      next();
    });

    server.get("*", (req, res) => {
      return handle(req, res);
    });

    server.listen(process.env.PORT || 3000, (err) => {
      if (err) throw err;
      console.log("> Ready on http://localhost:" + process.env.PORT);
    });
  })
  .catch((ex) => {
    console.error(ex.stack);
    process.exit(1);
  });

```

**Reply by ijjk** [reply]
Hi @dhavalsoni2001, can you provide a link to a repository with a minimal reproduction and exact steps to reproduce the behavior that was working with the custom server in Next.js v8 that is no longer working in v9? It's pretty difficult to investigate from just snippets. With a link to a repository we could investigate much better!

**Comment by dhavalsoni2001** [comment, marked as answer]
Closing the thread as I have successfully migrated the version.
